File: 2021/Day06.py
def simulate(input, days):
    counts = [input.count(i) for i in range(9)]
    for _ in range(days):
        counts = counts[1:] + counts[:1]
        counts[6] += counts[8]
    return sum(counts)

def part_1(input):
    return simulate(input, 80)

def part_2(input):
    return simulate(input, 256)

with open("2021/Day06_input.txt") as f:
    input = [int(x) for x in f.read().split(',')]

    print(f"Part 1: {part_1(input)}")
    print(f"Part 2: {part_2(input)}")

File: 2022/Day06.go
package main

import (
	"fmt"
	"math/bits"
	"os"
	"strings"
)

func marker(in string, n int) int {
	for i := n; i <= len(in); i++ {
		var m uint32
		for _, c := range in[i-n : i] {
			m |= 1 << (c - 'a')
		}
		if bits.OnesCount32(m) == n {
			return i
		}
	}
	return -1
}

func part1(in string) int { return marker(in, 4) }

func part2(in string) int { return marker(in, 14) }

func main() {
	b, _ := os.ReadFile("2022/Day06_input.txt")
	in := strings.TrimSpace(string(b))
	fmt.Println("Part 1:", part1(in))
	fmt.Println("Part 2:", part2(in))
}
